### DISCUSSION TITLE
bevy_qt plugin using cxx-qt

### DISCUSSION THREAD

**Discussion by drewpotter** [KDAB/cxx-qt#484] [Ideas]
I am currently interested in or attempting to use Qt/QML inside the wonderful Rust Bevy game engine.

I am planning to use the rendercontrol_opengl example which is actually included in the main qtdeclarative repo. I am thinking that it should be possible to achieve this.

If I succeed in this I will likely make a new MIT/Apache-v2 bevy plugin called bevy_qt, which will make use of your bindings.

Thanks to Andrew Hayzen for giving me some pointers on this. I think that the idea for passing information between Rust and C++ using QByteArray/QVector etc after creating the QML engine via Rust for now would be a great start.

In a future time, when we have the ability:
* to connect to signals on the Rust side
* to derive from a non-QObject constructed classes on the Rust side

then the plugin could evolve with those abilities.

I think this might be a good idea because Bevy is starting to take off and Qt is such a great library for UIs that I think it might be very worthwhile to get that working.

**Comment by Be-ing** [comment]
> to connect to signals on the Rust side

https://github.com/KDAB/cxx-qt/issues/473

> to derive from a non-QObject constructed classes on the Rust side

This is already possible. Anything inside the `#[cxx_qt::bridge]` module that doesn't have CXX-Qt specific annotations is forwarded to CXX. During development, we have used this to write some C++ code, call it from Rust, then later add features to our code generators and cxx-qt-lib bindings so we don't need the C++ code anymore.

**Reply by ahayzen-kdab** [reply]
> > to derive from a non-QObject constructed classes on the Rust side
> 
> This is already possible. Anything inside the `#[cxx_qt::bridge]` module that doesn't have CXX-Qt specific annotations is forwarded to CXX.

I was meaning https://github.com/KDAB/cxx-qt/issues/474 eg if we wanted to derive from QQuick* elements in Rust we need a different parent in the constructor :-)

**Comment by ahayzen-kdab** [comment]
I think as a starting point we need to understand what bevy needs as an input ? Is this documented somewhere ?

Then try creating a basic bevy app with a Qt plugin that starts a the QML engine (we are missing QQuickWindow/QQuickView bindings at the moment, if you have a Window in QML this does the same but then we need to access these in C++ to do the framebuffer fun). Then using a C++ proxy see if any frames from the Qt side can be drawn into bevy.

Once that is successful then moving to pure Rust and optimising would be the next stages.

**Reply by drewpotter** [reply]
> I think as a starting point we need to understand what bevy needs as an input ? Is this documented somewhere ?

There is a great plugin I discovered which covers a similar concept: https://github.com/blaind/bevy_webview

Basically, a webview is being rendered to a texture in bevy. I am planning to study how this plugin is working to build something similar for Qt.

> Then try creating a basic bevy app with a Qt plugin that starts a the QML engine (we are missing QQuickWindow/QQuickView bindings at the moment, if you have a Window in QML this does the same but then we need to access these in C++ to do the framebuffer fun). Then using a C++ proxy see if any frames from the Qt side can be drawn into bevy.

Yes exactly.

> Once that is successful then moving to pure Rust and optimising would be the next stages.

🎉

**Reply by CarlSchwan** [reply]
Also here is another example for a bevy integration https://github.com/linebender/vello/blob/main/examples/with_bevy/src/main.rs

**Comment by drewpotter** [comment]
I found where the textures are handled: https://github.com/blaind/bevy_webview/blob/main/src/systems/webview_update_textures.rs
